url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/handlers/MessageHandler.java

```diff
@@ -0,0 +1,238 @@
+        JsonObject jobj = new JsonObject();
+        jobj.addProperty("log", message);
+        String value  = jobj.toString();
+        return queue.offer(message);
+    }
+    
+    //Send a broadcast message
+    public boolean broadcastMessage(String message) {
+        JsonObject jobj = new JsonObject();
+        jobj.addProperty("broadcast", message);
+        String value = jobj.toString();
+        return queue.offer(message);
+    }
+
+    //Test to make sure that connection is open
+    @Override
+    public boolean isConnected() {
+        if (connection.isOpen()) return true;
+        return false;
+    }
+
+    @Override
+    public void closeConnection() {
+        for (JSONListener c: listeners) {
+            c.connectionClosed();
+        }
+        connection.close();
+    }
+
+    
+    
+    
+    /*
+     * The connection class that actually creates and handles the websocket
+     * connection.  This is also where the registered listeners get their messages
+     * and where they get their connection information.
+     */
+    public class Connection extends WebSocketClient {
+    
+        public Connection(URI serverUri, Draft draft) {
+            super(serverUri, draft);
+        }
+
+        public Connection(URI serverURI) {
+            super(serverURI);
+        }
+
+        //Let all registered listeners know that a connection established
+        @Override
+        public void onOpen(ServerHandshake handshakedata) {
+            System.out.println( "opened connection");
+            for (JSONListener c: listeners) {
+                c.connected(true);
+            }
+        }
+
+        @Override
+        public void onMessage(String message) {
+            for (JSONListener c : listeners) {
+                c.messageReceived(message);
+            }
+        }
+
+        @Override
+        public void onClose(int code, String reason, boolean remote) {
+            for (JSONListener c: listeners) {
+                c.connectionClosed();
+            }
+            System.out.println( "Connection closed by " + ( remote ? "remote peer" : "me" ) );
+        }
+
+        @Override
+        public void onError(Exception ex) {
+            ex.printStackTrace();
+        }
+    }
+    
+    //Start the thread that locks on the queue.  Started here for safety
+    @Override
+    public void run() {
+        Thread t = new Thread(new QueueManager());
+        t.start();
+    }
+    
+    //A separate thread the removes messages from the queue and sends them to the connection
+    private class QueueManager implements Runnable {
+        @Override
+        public void run() {
+            try {
+                openConn();//Blocks
+            } catch (InterruptedException ex) {
+                Logger.getLogger(MessageHandler.class.getName()).log(Level.SEVERE, null, ex);
+            }
+            
+            while (true) {
+                try {
+                    String message = queue.take();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:** 
The loop containing this operation may iterate indefinitely if the operation throws an exception on multiple instances of its evaluation. As the operation throws an exception, it bypasses exit points in the loop and continues to the next iteration through a catch clause.

**Fix:** 
Add a retry counter to ensure that the loop terminates after a predefined number of iterations.